### PR TITLE
Overhaul formatting of output values

### PR DESCRIPTION
**Pull request by aswan**
This is a step toward consolidating logic about formatting output
values into the individual ZNG Type classes.  The Type.StringOf()
method gains an additional parameter to specify the output format
and various ad-hoc formatting code is now moved into StringOf()
implementations as appropriate.

One small functional change is that valid UTF-8 values >128 inside a
bstring are no longer escaped in zng output (they are still escaped by
default in zeek logs or optionally left alone with the -U command line
argument).

Unescaping of string/bstring during input is not yet done, but in the
interest of keeping the PR from getting too large, that weill happen
in a separate follow-up.

### CODE REVIEW

File: zng/type.go [hunk 1]
```diff
@@ -29,16 +29,35 @@ type Resolver interface {
...
+// values, these are used to inform the formatter how containers should be
+// encoded and what sort of escaping should be applied to string types.
+type OutFmt int
+
+const (
+	OUT_FORMAT_UNESCAPED = iota
+	OUT_FORMAT_ZNG
+	OUT_FORMAT_ZEEK
+	OUT_FORMAT_ZEEK_ASCII
+	OUT_FORMAT_DEBUG
```
Comment (nwt) [inline zng/type.go]: The Go convention is to use mixed caps, even for constants.

File: zng/value.go [hunk 2]
```diff
@@ -71,22 +71,19 @@ func badZng(err error, t Type, zv zcode.Bytes) string {
 	return fmt.Sprintf("<ZNG-ERR type %s [%s]: %s>", t, zv, err)
 }
 
-// String implements the fmt.Stringer interfae and returns the string representation
-// of the value in accordance with the ZNG spec.  XXX currently we're not doing
-// escaping here.
-func (v Value) String() string {
+func (v Value) FormatAs(fmt OutFmt) string {
```
Comment (nwt) [inline zng/value.go]: I think we should just call this Format.

File: zng/vector.go [hunk 3]
```diff
@@ -39,22 +40,47 @@ func (t *TypeVector) Parse(in []byte) (zcode.Bytes, error) {
 	panic("zeek.TypeVector.Parse shouldn't be called")
 }
 
-func (t *TypeVector) StringOf(zv zcode.Bytes) string {
-	s := "vector["
-	comma := ""
+func (t *TypeVector) StringOf(zv zcode.Bytes, fmt OutFmt) string {
+	if len(zv) == 0 && fmt == OUT_FORMAT_ZEEK || fmt == OUT_FORMAT_ZEEK_ASCII {
```
Comment (nwt) [inline zng/vector.go]: I think you want parentheses around the || expression.

File: zng/value.go [hunk 4]
```diff
@@ -71,22 +71,19 @@ func badZng(err error, t Type, zv zcode.Bytes) string {
...
+	return v.Type.StringOf(v.Bytes, fmt)
 }
 
-// Format tranforms a zval encoding with its type encoding to a
-// a human-readable (and zng text-compliant) string format
-// encoded as a byte slice.
-//XXX this could be more efficient
-func (v Value) Format() []byte {
-	return []byte(Escape([]byte(v.String())))
+// String implements the fmt.Stringer interfae and returns a
```
Comment (nwt) [inline zng/value.go]: s/interfae/interface/

File: zng/vector.go (hunk 3 above)
Comment (aswan) [reply zng/vector.go]: thanks for catching that (but my thorough tests would have caught it eventually 😮 )

File: zng/string.go [hunk 5]
```diff
@@ -35,17 +41,45 @@ func (t *TypeOfString) String() string {
...
+	var s string
+	if len(code) == 4 {
+		s = fmt.Sprintf("\\u%s", code)
+	} else {
+		s = fmt.Sprintf("\\u{%s}", code)
+	}
+	return []byte(s)
+}
+
+func (t *TypeOfString) StringOf(zv zcode.Bytes, ofmt OutFmt) string {
```
Comment (henridf) [inline zng/string.go]: mininit: name use `fmt` instead of `ofmt` for consistency

Comment (henridf) [reply zng/string.go]: (or ofmt in the other `StringOf`s

Comment (aswan) [reply zng/string.go]: Ah, in an earlier draft I was using the golang fmt module and the names collided but that's no longer the case, will fix.

**Pull request merged**